### DISCUSSION TITLE
parallel error in HPC

### DISCUSSION THREAD

**Discussion by eric199405** [idaholab/moose#28715] [Q&A General] [answered]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

Dear MOOSE team,

I have installed moose in the hpc at Singapore (NSCC), using the conda framework.

However, when I submit a parallel case, actually it doesn't work.

```
Parallelism:
  Num Processors:          1
  Num Threads:             1
```

My qsub file is as follow.

```
#!/bin/bash
#PBS -q normal
#PBS -j oe
#PBS -N p12_nognd
#PBS -l select=1:ncpus=32:mpiprocs=32:mem=100GB
#PBS -l walltime=24:00:00

conda activate moose

  srun -n 32 /scratch/users/nus/guozixu/scut/scut-opt -i /scratch/users/nus/guozixu/scut/run.i > /scratch/users/nus/guozixu/scut/p12_nognd.txt
```
Could you give me some suggestions to address this issue?

**Comment by GiudGiud** [comment]
Hello

First try this
```
 srun -n 32 hostname
```
to make sure 32 processes are launched.
If so, then it's likely that the mpirun used under the hood is from a different MPI distribution than the one you used to compile moose.

you can use
```
which mpicc
```
when you compile and
```
which mpirun
```
when you run to compare

**Reply by eric199405** [reply]
Hi GiudGiud,
Thanks for your reply,
When I run `mpirun -n 32 hostname`，it shows that asp2a-login-nus01
```
asp2a-login-nus01
asp2a-login-nus01
asp2a-login-nus01
asp2a-login-nus01
asp2a-login-nus01
asp2a-login-nus01
asp2a-login-nus01
asp2a-login-nus01
asp2a-login-nus01
asp2a-login-nus01
asp2a-login-nus01
asp2a-login-nus01
asp2a-login-nus01
asp2a-login-nus01
asp2a-login-nus01
asp2a-login-nus01
asp2a-login-nus01
asp2a-login-nus01
asp2a-login-nus01
asp2a-login-nus01
asp2a-login-nus01
asp2a-login-nus01
asp2a-login-nus01
asp2a-login-nus01
asp2a-login-nus01
asp2a-login-nus01
asp2a-login-nus01
asp2a-login-nus01
asp2a-login-nus01
asp2a-login-nus01
asp2a-login-nus01

```
if I run which mpicc and which mpirun, it shows that 
`~/.conda/envs/moose/bin/mpicc
~/.conda/envs/moose/bin/mpirun`.

Kindly note that the parallel file I used is 
```
#!/bin/bash
#PBS -q normal
#PBS -j oe
#PBS -N p12_nognd
#PBS -l select=1:ncpus=32:mpiprocs=32:mem=100GB
#PBS -l walltime=24:00:00
# cd /scratch/users/nus/guozixu/scut/
# module load openmpi/4.1.2-hpe
conda activate moose

  mpirun -n 32 hostname

  mpirun -n 32 /scratch/users/nus/guozixu/scut/scut-opt -i /scratch/users/nus/guozixu/scut/run.i > /scratch/users/nus/guozixu/scut/p12_nognd.txt

```
The issue still exist.

**Reply by GiudGiud** [reply]
Hello

There are a few issues here

Why are you using conda compilers on a computing cluster? This won't scale well. 

Then after compiling with the MPI from conda, you are likely using MPI from a module to run. 

which mpirun should be run inside a job to see what is being used then

**Reply by eric199405** [reply]
Thanks for the reply.

I use conda to install MOOSE in HPC because the module ava in our HPC cannot support the normal installation. 

I have changed the job file:

```
#!/bin/bash
#PBS -q normal
#PBS -j oe
#PBS -N p12_nognd
#PBS -l select=1:ncpus=32:mpiprocs=32:mem=100GB
#PBS -l walltime=24:00:00

conda init --all

conda activate moose

  mpirun -n 32 hostname
  
  which mpirun

  mpirun -n 32 /scratch/users/nus/guozixu/scut/scut-opt -i /scratch/users/nus/guozixu/scut/run.i > /scratch/users/nus/guozixu/scut/p12_nognd.txt
```
and the output is 
```
CondaError: Run 'conda init' before 'conda activate'

x1001c2s0b0n0
x1001c2s0b0n0
x1001c2s0b0n0
x1001c2s0b0n0
x1001c2s0b0n0
x1001c2s0b0n0
x1001c2s0b0n0
x1001c2s0b0n0
x1001c2s0b0n0
x1001c2s0b0n0
x1001c2s0b0n0
x1001c2s0b0n0
x1001c2s0b0n0
x1001c2s0b0n0
x1001c2s0b0n0
x1001c2s0b0n0
x1001c2s0b0n0
x1001c2s0b0n0
x1001c2s0b0n0
x1001c2s0b0n0
x1001c2s0b0n0
x1001c2s0b0n0
x1001c2s0b0n0
x1001c2s0b0n0
x1001c2s0b0n0
x1001c2s0b0n0
x1001c2s0b0n0
x1001c2s0b0n0
x1001c2s0b0n0
x1001c2s0b0n0
x1001c2s0b0n0
x1001c2s0b0n0
/opt/cray/pe/pals/1.1.6/bin/mpirun
```
seems that the '/opt/cray/pe/pals/1.1.6/bin/mpirun' differs with '~/.conda/envs/moose/bin/mpicc ~/.conda/envs/moose/bin/mpirun'.

Is there any suggestions?

**Reply by milljm** [reply, marked as answer]
Can you try to activate the environment using an alternative method:

```bash
#!/bin/bash
#PBS -q normal
#PBS -j oe
#PBS -N p12_nognd
#PBS -l select=1:ncpus=32:mpiprocs=32:mem=100GB
#PBS -l walltime=24:00:00


# alternative way to 'activate' a conda environment
# see tip below why I don't use ~/
export PATH=/the/full/path/to/conda/bin:$PATH
source activate /the/full/path/to/.conda/envs/moose

which mpirun

srun -n 32 /scratch/users/nus/guozixu/scut/scut-opt \
  -i /scratch/users/nus/guozixu/scut/run.i \
  > /scratch/users/nus/guozixu/scut/p12_nognd.txt
```
> [!TIP]
> use `/full/path/to/**` paths instead of `~/`.
> I've seen shorthand path to home, fail in job submission scripts. Best to use fully declared paths in scripts like these.


I think `srun` is a SLURM job scheduler submitter. But I see PBS job scheduler attributes here (commonly used with `qsub`). Just throwing that out there. Not sure what is causing your error at the moment. But if you would try to activate Conda using the method above and try again/report back...

And for completeness, here would be the SLURM equivalent to your above PBS submission script:
```bash
#SBATCH --time=24:00:00      # walltime
#SBATCH --ntasks-per-node=32 # number of processor cores (i.e. tasks)
#SBATCH --nodes=1            # number of nodes
#SBATCH -J "p12_nognd"       # job name
#SBATCH -p normal            # queue
```

**Reply by eric199405** [reply]
It works! Special thanks for your helps!

**Reply by milljm** [reply]
Welcome! We activate our Conda environments in the same manner on our buildfarm (https://civet.inl.gov). I find it's a cleaner, less intrusive way that does not modify nor rely on, your `~/.bashrc, ~/.bash_profile, etc` profiles in order to work. I suspect when your SLURM job launches, it's probably ignoring some aspects of your shell profile (where/when Conda initializes).

**Reply by eric199405** [reply]
Thanks~
May I ask if use conda-based MOOSE in HPC, is there any potential computation efficiency lose?

**Reply by GiudGiud** [reply]
There is a performance issue. The inter-node communication will not use any optimized communication network your HPC cluster team deployed (infiniband for example)

On a single node, you should be fine. It will still be slower because the architecture-specific optimizations are not turned on

**Reply by eric199405** [reply]
Correct! 
I have tried 2 nodes, and it's very very slow.

**Reply by GiudGiud** [reply]
then you should work with your cluster team to install the correct way, using the MPI distribution they installed for you
https://mooseframework.inl.gov/getting_started/installation/hpc_install_moose.html

**Reply by eric199405** [reply]
Hi teams,
After I successfully submit the job to HPC, I run it for 1 day, then crash happens. 

The error code is

`libMesh terminatinglibMesh terminatinglibMesh terminatinglibMesh terminatinglibMesh terminatinglibMesh terminatinglibMesh terminatinglibMesh terminatinglibMesh terminatinglibMesh terminatinglibMesh terminatinglibMesh terminatinglibMesh terminatinglibMesh terminatinglibMesh terminatinglibMesh terminatinglibMesh terminatinglibMesh terminatinglibMesh terminatinglibMesh terminatinglibMesh terminatinglibMesh terminatinglibMesh terminatinglibMesh terminatinglibMesh terminatinglibMesh terminatinglibMesh terminatinglibMesh terminatinglibMesh terminatinglibMesh terminatinglibMesh terminatinglibMesh terminatinglibMesh terminatinglibMesh terminatinglibMesh terminating:
filesystem error: cannot check if file is empty: No such file or directory [/scratch/users/nus/guozixu/scut/run_out_cp/0178-restart-89.rd]
libMesh terminatinglibMesh terminatinglibMesh terminating:
filesystem error: cannot check if file is empty: No such file or directory [/scratch/users/nus/guozixu/scut/run_out_cp/0178-restart-102.rd]libMesh terminatinglibMesh terminatinglibMesh terminatinglibMesh terminating:
filesystem error: cannot check if file is empty: No such file or directory [/scratch/users/nus/guozixu/scut/run_out_cp/0178-restart-115.rd]
libMesh terminating:
filesystem error: cannot check if file is empty: No such file or directory [/scratch/users/nus/guozixu/scut/run_out_cp/0178-restart-120.rd]
libMesh terminating:
filesystem error: cannot check if file is empty: No such file or directory [/scratch/users/nus/guozixu/scut/run_out_cp/0178-restart-121.rd]
libMesh terminating:
filesystem error: cannot check if file is empty: No such file or directory [/scratch/users/nus/guozixu/scut/run_out_cp/0178-restart-123.rd]`

My qsub file is 

`#!/bin/bash
#PBS -q normal
#PBS -j oe
#PBS -N p12_nognd
#PBS -l select=1:ncpus=128:mpiprocs=128:mem=150GB
#PBS -l walltime=24:00:00

export PATH=$HOME/miniforge/bin:$PATH

source activate /home/users/nus/guozixu/.conda/envs/moose

   mpirun -n 128 hostname
  
   which mpirun

   mpirun -n 128 /scratch/users/nus/guozixu/scut/scut-opt -i /scratch/users/nus/guozixu/scut/run.i > /scratch/users/nus/guozixu/scut/p12_nognd.txt`

My Executioner is 

`  type = Transient
  solve_type = 'PJFNK'
  petsc_options = '-snes_ksp_ew'
  petsc_options_iname = '-pc_type -pc_hypre_type -ksp_gmres_restart'
  petsc_options_value = 'hypre    boomeramg          51'
  line_search = 'none'`

Could you please give me some suggestions for this issue?

**Reply by GiudGiud** [reply]
Hello

It seems checkpoint output failed.

This is turned on by default so you don't loose the run progress when such a crash happens. Now you should be able to restart the simulation from right before it crashed using the restart_file_base parameter in the Problem 

Guillaume

**Reply by eric199405** [reply]
Hi team,

When I run restart like that

```
[Problem]
  type = FEProblem
  restart_file_base = /scratch/users/nus/guozixu/scut/run_p12_nognd_cp_cp/LATEST # restart from LATEST time
  allow_initial_conditions_with_restart = true
[]


[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
[]

[Mesh]
  type = FileMesh
  file = /scratch/users/nus/guozixu/scut/run_p12_nognd_cp_cp/LATEST
[]
```

The following error occurs:

```
[36m
*** Info ***
SolidMechanics Action: selecting 'incremental finite strain' formulation.[39m

[33m
*** Warning, This code is deprecated and will be removed in future versions:
/scratch/users/nus/guozixu/scut/run.i:1329: (Outputs/e_p12_nognd2/interval):
'interval' has been deprecated and will be removed on 02/01/2025. Please use 'time_step_interval' instead.
[39m
Currently Setting Up
  Setting Up Undisplaced Mesh
    Finished Preparing Mesh                                                              [[33m  3.64 s[39m] [[33m  523 MB[39m]
  Finished Setting Up Undisplaced Mesh                                                   [[33m  3.64 s[39m] [[33m  523 MB[39m]
  Currently Setting Up Displaced Mesh
    Finished Preparing Mesh                                                              [[33m  2.09 s[39m] [[33m  643 MB[39m]
  Finished Setting Up Displaced Mesh                                                     [[33m  2.09 s[39m] [[33m  643 MB[39m]

[36m
*** Info ***
Using /scratch/users/nus/guozixu/scut/run_p12_nognd_cp_cp/0784 for restart.[39m
  Initializing
    Finished Initializing Equation Systems                                               [[33m  4.65 s[39m] [[33m  831 MB[39m]
  Finished Initializing                                                                  [[33m 12.77 s[39m] [[33m  883 MB[39m]
Finished Setting Up                                                                      [[33m 25.00 s[39m] [[33m  883 MB[39m]
Framework Information:
MOOSE Version:           git commit 08164beadc on 2024-09-25
LibMesh Version:         
PETSc Version:           3.21.4
SLEPc Version:           3.21.1
Current Time:            Mon Sep 30 12:47:09 2024
Executable Timestamp:    Sat Sep 28 01:38:47 2024

Checkpoint:
  Wall Time Interval:      Disabled
  User Checkpoint:         Outputs/p12_nognd2
  # Checkpoints Kept:      2
  Execute On:              TIMESTEP_END 

Parallelism:
  Num Processors:          128
  Num Threads:             1

Mesh: 
  Parallel Type:           replicated
  Mesh Dimension:          3
  Spatial Dimension:       3
  Nodes:                   
    Total:                 321602
    Local:                 2660
    Min/Max/Avg:           2326/2718/2512
  Elems:                   
    Total:                 160000
    Local:                 1256
    Min/Max/Avg:           1220/1278/1250
  Num Subdomains:          3
  Num Partitions:          128
  Partitioner:             metis

Nonlinear System:
  Num DOFs:                964806
  Num Local DOFs:          7980
  Variables:               { "disp_x" "disp_y" "disp_z" } 
  Finite Element Types:    "LAGRANGE" 
  Approximation Orders:    "FIRST" 

Auxiliary System:
  Num DOFs:                39040000
  Num Local DOFs:          306464
  Variables:               { "gp_act" "damage1" "damage2" "ssd_norm" "gnd_norm" ... "slip_increment_vector_7" 
                             "slip_increment_vector_8" "slip_increment_vector_9" "slip_increment_vector_10" 
                             "slip_increment_vector_11" } { "dslip_increment_dedge_0" "dslip_increment_dedge_1" 
                             "dslip_increment_dedge_2" "dslip_increment_dedge_3" "dslip_increment_dedge_4" 
                             ... "slip_resistance_8" "slip_resistance_9" "slip_resistance_10" "slip_resistance_11" 
                             "slip_resistance_12" } { "e_xx" "e_xy" "e_xz" "e_yx" "e_yy" ... "e_zx" "e_zy" 
                             "e_zz" "gss1" "slip_increment" } { "stress_xx" "stress_xy" "stress_yy" "strain_xx" 
                             "strain_xy" "strain_zz" "strain_xz" "strain_yz" } 
  Finite Element Types:    "MONOMIAL" "MONOMIAL" "MONOMIAL" "MONOMIAL" 
  Approximation Orders:    "FIRST" "CONSTANT" "FIRST" "CONSTANT" 

Execution Information:
  Executioner:             Transient
  TimeStepper:             IterationAdaptiveDT
  TimeIntegrator:          ImplicitEuler
  Solver Mode:             Preconditioned JFNK
  PETSc Preconditioner:    hypre boomeramg strong_threshold: 0.7 (auto)


Time Step 0, time = 8.31351

Postprocessor Values:
+----------------+----------------+----------------+----------------+----------------+----------------+----------------+
| time           | ave_gss1       | ave_rho_gnd    | ave_rho_ssd    | damage1        | damage2        | e_xx           |
+----------------+----------------+----------------+----------------+----------------+----------------+----------------+
|   8.313514e+00 |   4.729881e+02 |   0.000000e+00 |   3.819980e+00 |   1.647536e+01 |   4.061593e-01 |   8.357579e-03 |
+----------------+----------------+----------------+----------------+----------------+----------------+----------------+
+----------------+----------------+----------------+----------------+----------------+----------------+----------------+
| time           | e_xx_matrix    | e_xx_p1        | e_xx_p2        | gp_act1        | gp_act2        | max_gss1       |
+----------------+----------------+----------------+----------------+----------------+----------------+----------------+
|   8.313514e+00 |   8.453188e-03 |   5.920640e-03 |   8.321762e-03 |   1.273125e-02 |   7.269678e-01 |   9.729360e+02 |
+----------------+----------------+----------------+----------------+----------------+----------------+----------------+
+----------------+----------------+----------------+----------------+------------------+----------------+----------------+
| time           | max_rho_gnd    | max_rho_ssd    | stress_xx      | stress_xx_matrix | stress_xx_p1   | stress_xx_p2   |
+----------------+----------------+----------------+----------------+------------------+----------------+----------------+
|   8.313514e+00 |   0.000000e+00 |   5.175198e+00 |   1.036602e+03 |     1.032870e+03 |   1.070854e+03 |   1.049797e+03 |
+----------------+----------------+----------------+----------------+------------------+----------------+----------------+

Currently Executing
  Outputting
    Finished Outputting Step                                                             [[33m  8.51 s[39m] [[33m  956 MB[39m]
  Finished Outputting                                                                    [[33m  8.52 s[39m] [[33m  956 MB[39m]

Time Step 1, time = 8.31352, dt = 1e-05

===================================================================================
=   BAD TERMINATION OF ONE OF YOUR APPLICATION PROCESSES
=   PID 642973 RUNNING AT x1001c1s6b0n0.hostmgmt2001.cm.asp2a.nscc.sg
=   EXIT CODE: 139
=   CLEANING UP REMAINING PROCESSES
=   YOU CAN IGNORE THE BELOW CLEANUP MESSAGES
===================================================================================
YOUR APPLICATION TERMINATED WITH THE EXIT STRING: Segmentation fault (signal 11)
This typically refers to a problem with your application.
Please see the FAQ page for debugging suggestions
```
My qsub file is that

```
#!/bin/bash
#PBS -q normal
#PBS -j oe
#PBS -N p12_nognd
#PBS -l select=1:ncpus=128:mpiprocs=128:mem=150GB
#PBS -l walltime=24:00:00

export PATH=$HOME/miniforge/bin:$PATH

source activate /home/users/nus/guozixu/.conda/envs/moose

   mpirun -n 128 hostname
  
   which mpirun

   mpirun -n 128 /scratch/users/nus/guozixu/scut/scut-opt -i /scratch/users/nus/guozixu/scut/run.i --restart > /scratch/users/nus/guozixu/scut/p12_nognd2.txt

```

Wish to see any suggestions from your team.

**Reply by GiudGiud** [reply]
can you try using the last but one checkpoint?
since the code crashed while writing the latest, it could be that the file is corrupted
